### ISSUE TITLE
AI unintentionally declares war (runs over units)

### ISSUE THREAD

**Issue by LynxAbraxas** [civctp2/civctp2#224]
Even with #101, I still observe that the AI attacks civilian units when not at war with the unit owner. I have the impression that the AI has ordered such a unit/army with something different than actually attacking, but when it encounters a civilian unit while going to the point of order, it attacks that unit unconditionally (similar to #91). 

Here e.g. a Sumarian tank's order is "Group to (2, 112), Seige (9, 112)" but on its way there it comes across a Persian lawyer and attacks it (eventhough it had a cease-fire agreement with the Persians):

On its way, north of Isin:
![ss_2019-11-28_23:06:13](https://user-images.githubusercontent.com/21012234/69917107-d7e9dd80-1462-11ea-8987-c715df15324f.png)

attacking Lawyer (north west of Kish):
![ss_2019-11-28_23:06:26](https://user-images.githubusercontent.com/21012234/69917103-d3bdc000-1462-11ea-9d13-439b67207925.png)

Order still unchanged ("Group to (2, 112), Seige (9, 112)") after killing Lawyer:
![ss_2019-11-28_23:06:45](https://user-images.githubusercontent.com/21012234/69917102-d3bdc000-1462-11ea-9a33-934230256015.png)

Similarly, I also observed that a unit/army of a human player given an order (that includes a go-to) attacks another civ's unit/army if encountered along the way. However, if that would lead to war, the human player is shown the message that asks if one want's to declare war. Since that message/question does not exist for the AI, this could possibly be the reason for the unintentional declaration of war by running-over.

**Comment by LynxAbraxas** [comment]
Getting this assert with the debug version for the attack of the lawyer:
```
Assertion (!g_player[owner]->IsRobot() || Diplomat::GetDiplomat(owner).HasWarOrDesiresPreemptivelyWith(defender->GetOwner())) Failed in File:armyevent.cpp, Line:810
```

**Comment by LynxAbraxas** [comment]
Getting this assert with the debug version for the attack of the lawyer:
```
Assertion (!g_player[owner]->IsRobot() || Diplomat::GetDiplomat(owner).HasWarOrDesiresPreemptivelyWith(defender->GetOwner())) Failed in File:armyevent.cpp, Line:810
```
for  [dbgSue.zip](https://github.com/civctp2/civctp2/files/3936133/dbgSue.zip).

**Comment by LynxAbraxas** [comment]
My current impression is that this originates from
https://github.com/civctp2/civctp2/blob/623e59c962e78a8ff5eed763b02ef43977946fac/ctp2_code/gs/gameobj/ArmyData.cpp#L7227-L7235
because
https://github.com/civctp2/civctp2/blob/623e59c962e78a8ff5eed763b02ef43977946fac/ctp2_code/ai/diplomacy/diplomat.cpp#L5155-L5158
returns true.
So I suspect that the chosen conditions of
https://github.com/civctp2/civctp2/blob/623e59c962e78a8ff5eed763b02ef43977946fac/ctp2_code/ai/diplomacy/diplomat.cpp#L5172-L5243
are not sufficiently well balanced.
@MartinGuehmann Some of the related code changes seems to originate from you. Any idea how to improve this?

**Comment by LynxAbraxas** [comment]
It might be best to avoid getting to 
https://github.com/civctp2/civctp2/blob/623e59c962e78a8ff5eed763b02ef43977946fac/ctp2_code/gs/gameobj/ArmyData.cpp#L7227-L7235
by testing this condition already in 
https://github.com/civctp2/civctp2/blob/67954be9d07cfe5944dec129d0a5f21114a9682b/ctp2_code/gs/gameobj/armyevent.cpp#L812-L826
where currently only an assertion is based on that but not the follwoing code. @MartinGuehmann What do you think?

**Comment by LynxAbraxas** [comment]
Shouldn't
https://github.com/civctp2/civctp2/blob/67954be9d07cfe5944dec129d0a5f21114a9682b/ctp2_code/gs/gameobj/armyevent.cpp#L821-L825
come before
https://github.com/civctp2/civctp2/blob/67954be9d07cfe5944dec129d0a5f21114a9682b/ctp2_code/gs/gameobj/armyevent.cpp#L814-L819

**Comment by MartinGuehmann** [comment]
> Shouldn't
> https://github.com/civctp2/civctp2/blob/67954be9d07cfe5944dec129d0a5f21114a9682b/ctp2_code/gs/gameobj/armyevent.cpp#L821-L825
> 
> 
> come before
> https://github.com/civctp2/civctp2/blob/67954be9d07cfe5944dec129d0a5f21114a9682b/ctp2_code/gs/gameobj/armyevent.cpp#L814-L819
> 
> 
> ?
> It used to be different before [48d8afb](https://github.com/civctp2/civctp2/commit/48d8afb3a1531fbc5e522fa7e8003843aa9ae444):
> [48d8afb#diff-46145f8ebbe511caf828c937773ad5f4L818-R807](https://github.com/civctp2/civctp2/commit/48d8afb3a1531fbc5e522fa7e8003843aa9ae444#diff-46145f8ebbe511caf828c937773ad5f4L818-R807)

No, both are inserted after current, that means you insert FinishAttack first. After that it would be executed after the current event. But then you insert ClearOrders again after the current event. And so now ClearOrders is executed next and then comes FinishAttack.

Welcome to the world of asynchronous code execution. Have fun with this extra layer of complexity. And of course this is a source of bugs. :D

**Comment by LynxAbraxas** [comment]
Hm, OK so the execution is:
1. GEV_FinishAttack
2. do attack
3. GEV_ClearOrders

So what confuses me is `clears current oders via event before attacking` which I guess should be `clears current oders via event after attacking`.

**Comment by MartinGuehmann** [comment]
> Hm, OK so the execution is:
> 
> 1. GEV_FinishAttack
> 2. do attack
> 3. GEV_ClearOrders
> 
> So what confuses me is `clears current oders via event before attacking` which I guess should be `clears current oders via event after attacking`.

No, it is the other way around: First clear orders, for instance go to location X, then attack.

**Comment by MartinGuehmann** [comment]
This issue is fixed in f4cd3dae3e6a959833e401ee1c3ffa286fe9b1c9. Basically, if the AI bumps into a spy or another invisible unit. The unit is revealed and the orders of that army are cleared. However, the orders have been given. And the next army that bumps into that spy does not reveal it. And thus the AI runs over these units. In that case just clear orders and expel.
